src/session-pool.js

```diff
@@ -16,31 +16,50 @@
 
 'use strict';
 
-var events = require('events');
-var exec = require('methmeth');
+var EventEmitter = require('events').EventEmitter;
+var delay = require('delay');
 var extend = require('extend');
-var genericPool = require('generic-pool');
 var is = require('is');
+var PQueue = require('p-queue');
 var streamEvents = require('stream-events');
 var through = require('through2');
 var util = require('util');
 
+var DEFAULTS = {
+  acquireTimeout: Infinity,
+  concurrency: 10,
+  fail: false,
+  idlesAfter: 10,
+  keepAlive: 50,
+  max: 100,
+  maxIdle: 1,
+  min: 0,
+  writes: 0,
+};
+
+var READONLY = 'readonly';
+var READWRITE = 'readwrite';
+
 /**
  * Session pool configuration options.
  *
  * @typedef {object} SessionPoolOptions
- * @property {number} [acquireTimeout=0] Time in milliseconds before giving up
- *     trying to acquire a session. If the specified value is `0`, a timeout
- *     will not occur.
+ * @property {number} [acquireTimeout=Infinity] Time in milliseconds before
+ *     giving up trying to acquire a session. If the specified value is
+ *     `Infinity`, a timeout will not occur.
+ * @property {number} [concurrency=10] How many concurrent requests the pool is
+ *     allowed to make.
  * @property {boolean} [fail=false] If set to true, an error will be thrown when
  *     there are no available sessions for a request.
+ * @property {number} [idlesAfter=10] How long until a resource becomes idle, in
+ *     minutes.
  * @property {number} [max=100] Maximum number of resources to create at any
  *     given time.
  * @property {number} [maxIdle=1] Maximum number of idle resources to keep in
  *     the pool at any given time.
  * @property {number} [min=0] Minimum number of resources to keep in the pool at
  *     any given time.
- * @property {number} [keepAlive=59] How often to ping idle sessions, in
+ * @property {number} [keepAlive=50] How often to ping idle sessions, in
  *     minutes. Must be less than 1 hour.
  * @property {number} [writes=0] Pre-allocate transactions for the number of
```

Comment:
instead of fixed number of sessions, can this be made into a percentage of sessions they want to be prepared for write. That ways it will scale with how many open sessions there are.

Reply:
This is still pending.